src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -764,37 +763,25 @@ - (MTRAttributeValueWaiter *)waitForAttributeValues:(NSDictionary<MTRAttributePa
 
     if (attributeWaiter.allValuesSatisfied) {
         MTR_LOG("%@ waitForAttributeValues no need to wait, values already match: %@", self, values);
+        // We haven't added this waiter to self.attributeValueWaiters yet, so
+        // no need to remove it before notifying.
         [attributeWaiter _notifyWithError:nil];
         return attributeWaiter;
     }
 
-    // Otherwise, wait for the values to arrive or our timeout.
-    if (!self.attributeValueWaiters) {
-        self.attributeValueWaiters = [NSHashTable weakObjectsHashTable];
+    // Otherwise, wait for one of our termination conditions.  It's important to
+    // not take the lock earlier, because the _notifyWithError call above would
```

Comment:
This sounds outdated now since `_notifyWithError` doesn't go via the device lock anymore?

Reply:
Good catch, fixed.